File: packages/command-client/src/tauri/telemetry.ts
import type {
  TelemetryPreferences,
  TelemetrySendOutcome,
} from "../types";
import { call } from "./invoke";

export function getTelemetryPreferences() {
  return call<TelemetryPreferences>("telemetry_get_preferences");
}

export function setTelemetryEnabled(enabled: boolean) {
  return call<TelemetryPreferences>("telemetry_set_enabled", { enabled });
}

export function markTelemetryNoticeShown() {
  return call<TelemetryPreferences>("telemetry_mark_notice_shown");
}

export function recordTelemetryActive() {
  return call<TelemetrySendOutcome>("telemetry_record_active");
}

File: apps/desktop/src-tauri/src/telemetry.rs
use tauri::State;
use unfour_core::AppResult;
use unfour_telemetry::{
    TelemetryConfig, TelemetryPreferences, TelemetrySendOutcome, TelemetryService,
};

#[derive(Clone)]
pub struct TelemetryAppState {
    service: TelemetryService,
}

impl TelemetryAppState {
    pub fn new(service: TelemetryService) -> Self {
        Self { service }
    }
}

pub fn compiled_config() -> AppResult<TelemetryConfig> {
    let endpoint = match env!("UNFOUR_TELEMETRY_ENDPOINT") {
        "" => None,
        value => Some(value),
    };
    TelemetryConfig::for_current_target(
        env!("CARGO_PKG_VERSION"),
        env!("UNFOUR_RELEASE_CHANNEL"),
        env!("UNFOUR_DISTRIBUTION"),
        endpoint,
    )
}

#[tauri::command]
pub async fn telemetry_get_preferences(
    state: State<'_, TelemetryAppState>,
) -> AppResult<TelemetryPreferences> {
    state.service.preferences().await
}

#[tauri::command]
pub async fn telemetry_set_enabled(
    enabled: bool,
    state: State<'_, TelemetryAppState>,
) -> AppResult<TelemetryPreferences> {
    state.service.set_enabled(enabled).await
}

#[tauri::command]
pub async fn telemetry_mark_notice_shown(
    state: State<'_, TelemetryAppState>,
) -> AppResult<TelemetryPreferences> {
    state.service.mark_notice_shown().await
}

#[tauri::command]
pub async fn telemetry_record_active(
    state: State<'_, TelemetryAppState>,
) -> AppResult<TelemetrySendOutcome> {
    Ok(state.service.record_active().await)
}
